app/assets/javascripts/pretty-text/engines/discourse-markdown/code.js

```diff
@@ -1,35 +1,72 @@
 // we need a custom renderer for code blocks cause we have a slightly non compliant
 // format with special handling for text and so on
-
 const TEXT_CODE_CLASSES = ["text", "pre", "plain"];
 
-function render(tokens, idx, options, env, slf, md) {
-  let token = tokens[idx],
-    info = token.info ? md.utils.unescapeAll(token.info) : "",
-    langName = md.options.discourse.defaultCodeLang,
-    className,
-    escapedContent = md.utils.escapeHtml(token.content);
-
-  if (info) {
-    // strip off any additional languages
-    info = info.trim().split(/\s+/g)[0];
+function extractTokenInfo(info, md) {
+  if (!info) {
+    return;
+  }
+
+  info = md.utils.unescapeAll(info.replace(/[^\x00-\x7F]/g, "")).trim();
+
+  const matches = info.match(/(^\s*\S*)\s*(.*)/i);
+  if (!matches) {
+    return;
+  }
+
+  // ensure the token has only valid chars
+  if (!/^\w*$/i.test(matches[1])) {
+    return;
+  }
+
+  const extractedData = { tag: matches[1].trim(), attributes: {} };
+
+  if (matches[2]?.length) {
+    matches[2].split(",").forEach((potentialPair) => {
+      const [key, value] = potentialPair.trim().split(/\s+/g)[0].split("=");
+
+      // invalid pairs would get caught here and not used, eg `foo=`
+      if (key && value) {
+        extractedData.attributes[key] = value;
+      }
+    });
   }
 
-  const acceptableCodeClasses = md.options.discourse.acceptableCodeClasses;
-  if (
-    acceptableCodeClasses &&
-    info &&
-    acceptableCodeClasses.indexOf(info) !== -1
-  ) {
-    langName = info;
+  return extractedData;
+}
+
+function render(tokens, idx, options, env, slf, md) {
+  const token = tokens[idx];
+  const escapedContent = md.utils.escapeHtml(token.content);
+  const tokenInfo = extractTokenInfo(token.info, md);
+  const tag = tokenInfo?.tag || md.options.discourse.defaultCodeLang;
+  const attributes = tokenInfo?.attributes || {};
+
+  let className;
+
+  const acceptableCodeClasses =
+    md.options.discourse.acceptableCodeClasses || [];
+
+  if (TEXT_CODE_CLASSES.indexOf(tag) > -1) {
+    className = "lang-nohighlight";
+  } else if (acceptableCodeClasses.indexOf(tag) > -1) {
+    className = `lang-${tag}`;
+  } else {
+    className = "lang-nohighlight";
+    attributes["wrap"] = tag;
   }
 
-  className =
-    TEXT_CODE_CLASSES.indexOf(info) !== -1
-      ? "lang-nohighlight"
-      : "lang-" + langName;
+  const dataAttributes = Object.keys(attributes)
+    .map((key) => {
+      const value = md.utils.escapeHtml(attributes[key]);
+      key = md.utils.escapeHtml(key);
+      return `data-${key}="${value}"`;
+    })
```

Comment:
I think we'll need to either allowlist or namespace these data attributes. We don't want end-users to be able to generate things like `data-user-card` or `data-popover`.

(more info on previous issues like this at dev `/t/51197`)

Reply:
Yes, I don't think I can't do it in custom sadly, I tried to do it before.

I guess I can manage a list of custom exceptions for now?

Reply:
https://github.com/discourse/discourse/pull/15743/commits/44bd3e06b8b9b60e11761d57213de6c74ee19877